### DISCUSSION TITLE
[@types/argparse] npm packages out of sync: argparse and @types/argparse

### DISCUSSION THREAD

**Discussion by melonmouse** [DefinitelyTyped/DefinitelyTyped#63645] [Issues with a @types package]
**The symptoms**

Got an unexplained `TypeError: simpleExample.add_argument is not a function` when trying to use argparse in typescript (using an example from the docs).

**The cause**

The npm package `@types/argparse` defaulted to version `2.0.10` for me, while `argparse` defaulted to `argparse@1.0.10`. These are not compatible and cause errors for any of the functions that [had their names changed from v1 to v2](https://github.com/nodeca/argparse/blob/master/doc/migrate_v1_to_v2.md). 
Installing the newer version of argparse explicitly fixed it for me:
`npm install argparse@2`
This went wrong on npm versions 7.11.2 and 9.2.0.

**Possible fixes**

- upgrade the default of `argparse` to `2.0.10`
- downgrade [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) the default version of `@types/argparse` to 1.0.10
- update dependency requirements so npm either installs compatible versions or complains when installing incompatible versions

I also filed an issue in the main argparse github repo: https://github.com/nodeca/argparse/issues/172

**p.s.**
Thanks for adding types for so many packages! I love typescript and am thankful for your great work!

[a comment by an automated account is not shown]

**Comment by Semigradsky** [comment]
@melonmouse I not understand you, what means `argparse` defaulted to `argparse@1.0.10`?

If you used `argparse@1.x` than install `@types/argparse@1.x`. If you used `argparse@2.x` than install `@types/argparse@2.x`.

**Reply by usewits** [reply]
For me:

`npm install argparse`
`npm install @types/argparse`

Installed argparse 1.x and @types/argparse 2.x. I did not explicitly set the versions, but the result was not compatible. The defaults should be compatible.

**Reply by Semigradsky** [reply]
So you want to have `peerDependency` on `@types/argparse` package?
To have `ERESOLVE unable to resolve dependency tree` error when you try to install `@types/argparse` without specifying the correct version.

**Reply by Semigradsky** [reply]
For example, if you have installed react 17, than you will have this error after type `npm install react-dom`:
```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: tmp@1.0.0
npm ERR! Found: react@17.0.2
npm ERR! node_modules/react
npm ERR!   react@"^17.0.0" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer react@"^18.2.0" from react-dom@18.2.0
npm ERR! node_modules/react-dom
npm ERR!   react-dom@"*" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR!
npm ERR! For a full report see:
npm ERR! /home/dmitry/.npm/_logs/2022-12-20T19_11_45_295Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/dmitry/.npm/_logs/2022-12-20T19_11_45_295Z-debug-0.log
```

**Reply by Semigradsky** [reply]
As you see npm will not install react-dom 17 by default. It just try to install the latest version.

**Reply by usewits** [reply]
I'm no expert on npm package management, but your proposed solution sounds good to me.

**Reply by melonmouse** [reply]
Took a quick look and think this thread is not a duplicate. Moving it over to an issue. Let me know if that is appropriate (sorry if it isn't).

Thanks :D
